fairscale/nn/data_parallel/fully_sharded_data_parallel.py

```diff
@@ -857,7 +866,40 @@ def _register_hook(t: torch.Tensor) -> torch.Tensor:
         return outputs
 
     def _register_post_backward_hooks(self) -> None:
-        """Register backward hooks to reshard params and reduce-scatter grads."""
+        """
+        Register backward hooks to reshard params and reduce-scatter grads.
+
+        This is called during forward pass. The goal is to attach a hook
+        on each of the parameter's gradient generating function (``grad_acc``
+        below) so that the hook is called *after* all gradients for that
+        param are computed.
+
+        Goals:
+
+        1. We want the hook to fire once and only once *after* all gradients
+        are accumulated for a param.
+        2. If it fires more than once, we end up incorrectly shard the grad
+        multiple times. (could lead to dimension too small)
+        3. If it fires once but too early or doesn't fire, we leave gradients
+        unsharded. (could lead to dimension too large)
+
+        Due to multiple-pass forward, this function can be called on
+        the same parameter multiple times in a single forward pass. If we register
+        the hook multiple time, we end up getting called multiple times. We
+        could try to get a new hook every time and delete the previou one
+        registered. However, due to *unknown reason* (I have debugged it for
```

Comment:
This is not very satisfying since I didn't find out what's different with mixed precision. I did bunch of tests with and without FSDP, I couldn't reproduce the same behavior (different grad_acc objects) outside of FSDP. Within FSDP, nothing related to the param or the data within the param seems to trigger this. If someone else can figure this out, definitely help!

Reply:
Interesting! I'm not sure what's the cause. But there was something in the 1.8 release notes about backward hooks not behaving properly, which was fixed in 1.8. Search for "register_full_backward_hook" here: https://github.com/pytorch/pytorch/releases/tag/v1.8.0

Reply:
The new hook api is very interesting. I wonder what's the behavior will be like for the new hook function when we are doing nesting. We are not using the old API. I wonder if the way we do hooks is equivalent to the API there. I will do some digging.